File: api/providers/theoddsapi.py
"""TheOddsAPI provider implementation."""
from typing import List, Optional, Dict
from datetime import datetime
import httpx
from providers.base import BaseProvider, ProviderMatch, ProviderSnapshot
from services.odds_utils import (
    decimal_odds_to_implied_probability,
    normalize_probabilities_2way,
    normalize_probabilities_3way,
)
from config import settings
import logging

logger = logging.getLogger(__name__)


class TheOddsAPIProvider(BaseProvider):
    """
    TheOddsAPI provider.

    Docs: https://the-odds-api.com/liveapi/guides/v4/
    """

    BASE_URL = "https://api.the-odds-api.com/v4"

    # Sport keys mapping
    SPORT_KEYS = {
        "soccer": {
            "EPL": "soccer_epl",
        },
        "afl": {
            "AFL": "aussierules_afl",
        },
        "nrl": {
            "NRL": "rugbyleague_nrl",
        },
    }

    def __init__(self):
        super().__init__(name="TheOddsAPI", enabled=bool(settings.theoddsapi_key))
        self.api_key = settings.theoddsapi_key

    def _get_sport_key(self, sport: str, league: Optional[str] = None) -> Optional[str]:
        """Get TheOddsAPI sport key."""
        if sport not in self.SPORT_KEYS:
            return None

        if league:
            return self.SPORT_KEYS[sport].get(league)

        # Return first league for sport if no specific league requested
        return list(self.SPORT_KEYS[sport].values())[0] if self.SPORT_KEYS[sport] else None

    async def fetch_matches(
        self, sport: str, league: Optional[str] = None
    ) -> List[ProviderMatch]:
        """Fetch upcoming matches."""
        if not self.enabled or not self.api_key:
            self.logger.warning("Provider not enabled or API key missing")
            return []

        sport_key = self._get_sport_key(sport, league)
        if not sport_key:
            self.logger.warning(f"No sport key found for {sport}/{league}")
            return []

        url = f"{self.BASE_URL}/sports/{sport_key}/odds"
        params = {
            "apiKey": self.api_key,
            "regions": "au",  # Australian bookmakers
            "markets": "h2h",
            "oddsFormat": "decimal",
        }

        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.get(url, params=params)
                response.raise_for_status()

                data = response.json()
                matches = []

                for event in data:
                    match = ProviderMatch(
                        external_id=event["id"],
                        home_team=event["home_team"],
                        away_team=event["away_team"],
                        kickoff_time=datetime.fromisoformat(
                            event["commence_time"].replace("Z", "+00:00")
                        ),
                        league=league or sport,
                        sport=sport,
                    )
                    matches.append(match)

                self.logger.info(f"Fetched {len(matches)} matches for {sport_key}")
                return matches

        except Exception as e:
            self.logger.error(f"Error fetching matches: {e}")
            return []

    async def fetch_odds(
        self, match_external_id: str, sport: str
    ) -> Optional[ProviderSnapshot]:
        """Fetch current odds for a match."""
        if not self.enabled or not self.api_key:
            return None

        # Get sport key (need to know league, but we can try all)
        sport_keys = list(self.SPORT_KEYS.get(sport, {}).values())
        if not sport_keys:
            return None

        # Try each sport key
        for sport_key in sport_keys:
            url = f"{self.BASE_URL}/sports/{sport_key}/events/{match_external_id}/odds"
            params = {
                "apiKey": self.api_key,
                "regions": "au",
                "markets": "h2h",
                "oddsFormat": "decimal",
            }

            try:
                async with httpx.AsyncClient(timeout=30.0) as client:
                    response = await client.get(url, params=params)

                    if response.status_code == 404:
                        continue  # Try next sport key

                    response.raise_for_status()
                    data = response.json()

                    # Get first bookmaker with odds
                    if not data.get("bookmakers"):
                        continue

                    bookmaker = data["bookmakers"][0]
                    h2h_market = next(
                        (m for m in bookmaker["markets"] if m["key"] == "h2h"), None
                    )

                    if not h2h_market:
                        continue

                    # Extract odds
                    outcomes = {o["name"]: o["price"] for o in h2h_market["outcomes"]}

                    # Determine market type
                    market_type = self._convert_to_market_type(sport)

                    # Convert to probabilities
                    home_odds = outcomes.get(data["home_team"])
                    away_odds = outcomes.get(data["away_team"])

                    if not home_odds or not away_odds:
                        continue

                    home_prob = decimal_odds_to_implied_probability(home_odds)
                    away_prob = decimal_odds_to_implied_probability(away_odds)

                    # Check for draw
                    draw_odds = outcomes.get("Draw")
                    draw_prob = None

                    if draw_odds and market_type == "moneyline_3way":
                        draw_prob = decimal_odds_to_implied_probability(draw_odds)
                        home_prob, draw_prob, away_prob = normalize_probabilities_3way(
                            home_prob, draw_prob, away_prob
                        )
                    else:
                        home_prob, away_prob = normalize_probabilities_2way(
                            home_prob, away_prob
                        )

                    return ProviderSnapshot(
                        match_external_id=match_external_id,
                        captured_at=datetime.utcnow(),
                        market_type=market_type,
                        home_prob=home_prob,
                        away_prob=away_prob,
                        draw_prob=draw_prob,
                        raw_odds=outcomes,
                    )

            except Exception as e:
                self.logger.error(f"Error fetching odds for {match_external_id}: {e}")
                continue

        return None

    async def health_check(self) -> Dict:
        """Check provider health."""
        if not self.enabled or not self.api_key:
            return {
                "status": "disabled",
                "message": "Provider not enabled or API key missing",
                "latency_ms": 0,
            }

        url = f"{self.BASE_URL}/sports"
        params = {"apiKey": self.api_key}

        try:
            import time

            start = time.time()

            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.get(url, params=params)
                response.raise_for_status()

            latency_ms = (time.time() - start) * 1000

            return {
                "status": "healthy",
                "message": "API accessible",
                "latency_ms": round(latency_ms, 2),
            }

        except Exception as e:
            return {
                "status": "unhealthy",
                "message": str(e),
                "latency_ms": 0,
            }

File: web/src/lib/odds/providers/theoddsapi.ts
/**
 * TheOddsAPI provider client
 * Ported from Python theoddsapi.py
 */

import { TheOddsAPIEvent, ProviderOdds, RawBookmakerOdds } from '../types';
import {
  decimalOddsToImpliedProbability,
  normalizeOdds2Way,
  normalizeOdds3Way,
} from '../conversion';

const BASE_URL = 'https://api.the-odds-api.com/v4';

// Sport keys mapping - includes ALL sports for multi generator
const SPORT_KEYS: { [sport: string]: { [league: string]: string } } = {
  soccer: {
    EPL: 'soccer_epl',
    LA_LIGA: 'soccer_spain_la_liga',
    BUNDESLIGA: 'soccer_germany_bundesliga',
    SERIE_A: 'soccer_italy_serie_a',
    LIGUE_1: 'soccer_france_ligue_one',
    UEFA_CHAMPIONS: 'soccer_uefa_champs_league',
    UEFA_EUROPA: 'soccer_uefa_europa_league',
  },
  afl: {
    AFL: 'aussierules_afl',
  },
  nrl: {
    NRL: 'rugbyleague_nrl',
  },
  americanfootball: {
    NFL: 'americanfootball_nfl',
    NCAAF: 'americanfootball_ncaaf',
  },
  basketball: {
    NBA: 'basketball_nba',
    NCAAB: 'basketball_ncaab',
  },
  baseball: {
    MLB: 'baseball_mlb',
  },
  icehockey: {
    NHL: 'icehockey_nhl',
  },
  rugbyunion: {
    RUGBY: 'rugbyunion_super_rugby',
    RUGBY_UNION: 'rugbyunion_super_rugby',
  },
  rugbyleague: {
    RUGBY_LEAGUE: 'rugbyleague_nrl',
  },
  mma: {
    UFC: 'mma_mixed_martial_arts',
  },
  boxing: {
    BOXING: 'boxing_boxing',
  },
  // tennis: tennis_atp not available in AU region
  cricket: {
    CRICKET: 'cricket_test_match',
  },
  horse_racing: {
    RACING: 'horse_racing_australia',
  },
};

/**
 * Get TheOddsAPI sport key for a given sport and league
 */
function getSportKey(sport: string, league?: string): string | null {
  const sportLower = sport.toLowerCase();

  if (!SPORT_KEYS[sportLower]) {
    return null;
  }

  if (league) {
    return SPORT_KEYS[sportLower][league] || null;
  }

  // Return first league for sport if no specific league requested
  const leagues = Object.values(SPORT_KEYS[sportLower]);
  return leagues.length > 0 ? leagues[0] : null;
}

/**
 * Determine market type based on sport
 */
function getMarketType(sport: string): '2way' | '3way' {
  // Soccer has 3-way markets (home/draw/away)
  // AFL and NRL have 2-way markets (home/away)
  return sport.toLowerCase() === 'soccer' ? '3way' : '2way';
}

/**
 * TheOddsAPI client class
 */
export class TheOddsAPIClient {
  private apiKey: string;

  constructor(apiKey: string) {
    if (!apiKey) {
      throw new Error('TheOddsAPI key is required');
    }
    this.apiKey = apiKey;
  }

  /**
   * Fetch upcoming matches for a sport/league
   */
  async fetchMatches(sport: string, league?: string): Promise<TheOddsAPIEvent[]> {
    const sportKey = getSportKey(sport, league);
    if (!sportKey) {
      console.warn(`No sport key found for ${sport}/${league}`);
      return [];
    }

    const url = `${BASE_URL}/sports/${sportKey}/odds`;
    const params = new URLSearchParams({
      apiKey: this.apiKey,
      regions: 'au', // Australian bookmakers
      markets: 'h2h',
      oddsFormat: 'decimal',
    });

    try {
      const response = await fetch(`${url}?${params}`, {
        next: { revalidate: 300 }, // Cache for 5 minutes
      });

      if (!response.ok) {
        const errorText = await response.text();
        console.error(`TheOddsAPI error for ${sportKey}: ${response.status} ${response.statusText}`, errorText);
        throw new Error(`TheOddsAPI error (${sportKey}): ${response.status} ${response.statusText} - ${errorText}`);
      }

      const data = await response.json();
      console.log(`TheOddsAPI returned ${data.length} events for ${sportKey}`);
      return data as TheOddsAPIEvent[];
    } catch (error) {
      console.error(`Error fetching matches from TheOddsAPI for ${sportKey}:`, error);
      throw error;
    }
  }

  /**
   * Fetch current odds for a specific match
   */
  async fetchMatchOdds(
    eventId: string,
    sport: string,
    league?: string
  ): Promise<TheOddsAPIEvent | null> {
    const sportKey = getSportKey(sport, league);
    if (!sportKey) {
      console.warn(`No sport key found for ${sport}/${league}`);
      return null;
    }

    const url = `${BASE_URL}/sports/${sportKey}/events/${eventId}/odds`;
    const params = new URLSearchParams({
      apiKey: this.apiKey,
      regions: 'au',
      markets: 'h2h',
      oddsFormat: 'decimal',
    });

    try {
      const response = await fetch(`${url}?${params}`, {
        next: { revalidate: 60 }, // Cache for 1 minute
      });

      if (!response.ok) {
        if (response.status === 404) {
          return null;
        }
        throw new Error(`TheOddsAPI error: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      return data as TheOddsAPIEvent;
    } catch (error) {
      console.error(`Error fetching odds for event ${eventId}:`, error);
      throw error;
    }
  }

  /**
   * Extract and normalize odds from all bookmakers for an event
   */
  extractBookmakerOdds(event: TheOddsAPIEvent): RawBookmakerOdds[] {
    const odds: RawBookmakerOdds[] = [];

    for (const bookmaker of event.bookmakers) {
      const h2hMarket = bookmaker.markets.find(m => m.key === 'h2h');
      if (!h2hMarket) {
        continue;
      }

      // Extract outcomes
      const outcomes: { [name: string]: number } = {};
      for (const outcome of h2hMarket.outcomes) {
        outcomes[outcome.name] = outcome.price;
      }

      const homeOdds = outcomes[event.home_team];
      const awayOdds = outcomes[event.away_team];
      const drawOdds = outcomes['Draw'];

      if (!homeOdds || !awayOdds) {
        continue;
      }

      odds.push({
        bookmaker: bookmaker.title,
        home: homeOdds,
        away: awayOdds,
        draw: drawOdds,
        last_update: bookmaker.last_update,
      });
    }

    return odds;
  }

  /**
   * Convert bookmaker odds to provider odds with normalized probabilities
   */
  convertToProviderOdds(
    bookmakerOdds: RawBookmakerOdds[],
    marketType: '2way' | '3way'
  ): ProviderOdds[] {
    return bookmakerOdds.map(bo => {
      // Convert to probabilities
      const homeProb = decimalOddsToImpliedProbability(bo.home);
      const awayProb = decimalOddsToImpliedProbability(bo.away);
      const drawProb = bo.draw ? decimalOddsToImpliedProbability(bo.draw) : undefined;

      // Normalize
      let normalized: {
        home: number;
        away: number;
        draw?: number;
      };

      if (marketType === '3way' && drawProb !== undefined) {
        const [home, draw, away] = normalizeOdds3Way(homeProb, drawProb, awayProb);
        normalized = { home, draw, away };
      } else {
        const [home, away] = normalizeOdds2Way(homeProb, awayProb);
        normalized = { home, away };
      }

      return {
        provider: bo.bookmaker,
        odds: {
          home: bo.home,
          away: bo.away,
          draw: bo.draw,
        },
        probabilities: normalized,
        timestamp: bo.last_update,
      };
    });
  }

  /**
   * Fetch and aggregate odds for a match
   * Returns provider odds ready for aggregation
   */
  async fetchAndExtractOdds(
    eventId: string,
    sport: string,
    league?: string
  ): Promise<ProviderOdds[]> {
    const event = await this.fetchMatchOdds(eventId, sport, league);
    if (!event) {
      return [];
    }

    const bookmakerOdds = this.extractBookmakerOdds(event);
    const marketType = getMarketType(sport);

    return this.convertToProviderOdds(bookmakerOdds, marketType);
  }

  /**
   * Health check - verify API is accessible
   */
  async healthCheck(): Promise<{
    status: 'healthy' | 'unhealthy';
    message: string;
    latency_ms: number;
  }> {
    const url = `${BASE_URL}/sports`;
    const params = new URLSearchParams({
      apiKey: this.apiKey,
    });

    const start = Date.now();

    try {
      const response = await fetch(`${url}?${params}`, {
        next: { revalidate: 0 }, // No cache for health checks
      });

      if (!response.ok) {
        return {
          status: 'unhealthy',
          message: `API returned ${response.status}`,
          latency_ms: Date.now() - start,
        };
      }

      return {
        status: 'healthy',
        message: 'API accessible',
        latency_ms: Date.now() - start,
      };
    } catch (error) {
      return {
        status: 'unhealthy',
        message: error instanceof Error ? error.message : 'Unknown error',
        latency_ms: Date.now() - start,
      };
    }
  }
}

/**
 * Get TheOddsAPI client instance
 * Reads API key from environment variables
 */
export function getTheOddsAPIClient(): TheOddsAPIClient {
  const apiKey = process.env.THEODDSAPI_KEY;
  if (!apiKey) {
    throw new Error('THEODDSAPI_KEY environment variable is not set');
  }
  return new TheOddsAPIClient(apiKey);
}
